File: z_lc/11-container-with-most-water/container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        res = 0
        i, j = 0, len(height) - 1
        while i < j:
            h = min(height[i], height[j])
            area = h * (j - i)
            
            res = max(res, area)
            
            while i < j and height[i] < h:
                i += 1
            while i < j and height[j] < h:
                j -= 1
        return res

File: z_lc/11-container-with-most-water/container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        #include <algorithm>
        
        int res = 0;
        int i = 0, j = height.size() - 1;
        while (i < j) {
            int h = std::min(height[i], height[j]);
            int area = h * (j - i);
            
            res = std::max(res, area);
            
            while (i < j && height[i] <= h) {
                ++i;
            }
            while (i < j && height[j] <= h) {
                --j;
            }
        }
        return res;
    }
};